### PR TITLE
Switch alllocator to use a "freelist of segments"

### PR DESCRIPTION
**Pull request by bollu**
We change the freelist in the mblockallocator to become a list of "free segments". This eliminates error-prone bounds checking, and should be a more stable base for the megablock allocator. This should fix `CmpWord8`

### CODE REVIEW

File: asterius/rts/rts.mblockalloc.mjs [hunk 1]
```diff
@@ -1,13 +1,25 @@
 import * as rtsConstants from "./rts.constants.mjs";
 import { Memory } from "./rts.memory.mjs";
+import * as assert from "assert";
```
Comment (TerrorJack) [inline asterius/rts/rts.mblockalloc.mjs]: This is a node-only import which will break the browser target again. We can have a local no-op assert function here for now, before we figure out how to properly do rts assertions.

File: asterius/rts/rts.mblockalloc.mjs [hunk 2]
```diff
@@ -65,25 +85,26 @@ export class MBlockAlloc {
...
-      const bd = l_end + rtsConstants.offset_first_bdescr,
-            start = l_end + rtsConstants.offset_first_block,
-            blocks = (r - start) / rtsConstants.block_size;
-      this.memory.i64Store(bd + rtsConstants.offset_bdescr_start, start);
-      this.memory.i64Store(bd + rtsConstants.offset_bdescr_free, start);
-      this.memory.i64Store(bd + rtsConstants.offset_bdescr_link, 0);
-      this.memory.i32Store(bd + rtsConstants.offset_bdescr_blocks, blocks);
-      this.freeList.push(bd);
+        // memset a custom number purely for debugging help.
+        this.memory.memset(l_end, 0x42 + i, r - l_end);
```
Comment (TerrorJack) [inline asterius/rts/rts.mblockalloc.mjs]: Better to set to a larger byte e.g. 0xff so it's more likely to get an overflowing address.

File: asterius/rts/rts.mblockalloc.mjs [hunk 3]
```diff
@@ -30,31 +42,39 @@ export class MBlockAlloc {
...
-    const req_blocks = ((rtsConstants.mblock_size * n) - rtsConstants.offset_first_block) / rtsConstants.block_size;
-    for (let i = 0; i < this.freeList.length; ++i) {
-      const bd = this.freeList[i],
-            blocks = this.memory.i32Load(bd + rtsConstants.offset_bdescr_blocks);
+    const req_blocks = 
+          Math.ceil(((rtsConstants.mblock_size * n) - rtsConstants.offset_first_block) / 
+          rtsConstants.block_size);
+
+    for (let i = 0; i < this.freeSegments.length; ++i) {
+      const [bd, r] = this.freeSegments[i];
```
Comment (TerrorJack) [inline asterius/rts/rts.mblockalloc.mjs]: So, we are saving the (bd, right) tuple as a segment here, which in fact is not too far from the original design which only stores bd. Further more, the freeSegment function below still initializes the bdescr of freed space.

For simplicity, I think we should:

* Store (left, right) tuple, ensure left and right satisfy mblock alignment.
* freeSegment only takes care of wiping bytes.
* When allocMegaGroup picks up a tuple, it initializes the bdescr fields before returning.

File: asterius/rts/rts.mblockalloc.mjs (hunk 2 above)
Comment (bollu) [reply asterius/rts/rts.mblockalloc.mjs]: If we memset `0xff + i`, then we can't eyeball the bit-pattern to figure out which segment it came from, unfortunately.

**Pull request closed without merging**